docs: Physical layout docs improvements (#2533)
zmkfirmware/zmk: docs/docs/development/hardware-integration/physical-layouts.md

---
title: Physical Layouts
---

A physical layout is a devicetree entity that aggregates all details about a certain possible keyboard layout.
It contains:

- A [keyboard scan (kscan) driver](../../config/kscan.md)
- A [matrix transform](../../config/layout.md#matrix-transform)
- (Optional) [Physical key positions](#physical-layout-positions)

## Basic Physical Layout

A basic physical layout without the `keys` property looks like this:

```dts
/ {
    default_layout: default_layout {
        compatible = "zmk,physical-layout";
        display-name = "Default Layout";
        transform = <&default_transform>;
        kscan = <&kscan0>;
    };
};
```

> It is given a name, a matrix transform, and a kscan. If all of your physical layouts share the same kscan, then the `kscan` property can be omitted - in this case it needs to be set in the [`chosen` node](./new-shield.mdx#chosen-node). See the [configuration section on physical layouts](../../config/) for reference.

Reviewer: ```suggestion
It is given a name, a matrix transform, and a kscan. If all of your physical layouts share the same kscan, then the `kscan` property can be omitted - in this case it needs to be set in the [`chosen` node](./new-shield.mdx#chosen-node). See the [configuration section on physical layouts](../../config/index.md) for reference.
```

:::tip
You can specify negative values in devicetree using parentheses around it, e.g. `(-3000)` for a 30 degree counterclockwise rotation.
:::

ZMK defines a number of popular physical layouts in-tree at [`app/dts/layouts`](https://github.com/zmkfirmware/zmk/tree/main/app/dts/layouts).
To use such layouts, import them and assign their `transform` and (optionally) `kscan` properties.

### Physical Layout with Keys Example

Here is an example of a physical layout for a 2x2 macropad:

```dts
#include <physical_layouts.dtsi>

/ {
    macropad_physical_layout: macropad_physical_layout {
        compatible = "zmk,physical-layout";
        display-name = "Macro Pad";
        transform = <&default_transform>;
        kscan = <&kscan0>;
        keys  //                     w   h    x    y     rot    rx    ry
            = <&key_physical_attrs 100 100    0    0       0     0     0>
            , <&key_physical_attrs 100 100  100    0       0     0     0>
            , <&key_physical_attrs 100 100    0  100       0     0     0>
            , <&key_physical_attrs 100 100  100  100       0     0     0>
            ;
    };
};
```

> ### Predefined Layout Usage Example

Reviewer: ```suggestion
## Using Predefined Layouts
```
How about moving line 61 here and making this its own level 2 header?

As it is, it seemed a bit awkward to me to have this example with 4 layouts under "basic", *then* have a section on "multiple layouts"

Author: Good idea. I've also added a quick example using the ferris' layout, since that's not a multiple layouts situation, and a sentence to make the situation there clearer.

Here is an example of using the predefined physical layouts for a 60% keyboard:

```dts

> 

Reviewer: ```suggestion
```

#include <layouts/common/60percent/all1u.dtsi>
#include <layouts/common/60percent/ansi.dtsi>
#include <layouts/common/60percent/hhkb.dtsi>
#include <layouts/common/60percent/iso.dtsi>

> \\ Assigning the suitable matrix transforms

Reviewer: ```suggestion
// Assigning suitable kscan and matrix transforms
```

&layout_60_ansi {
    transform = <&ansi_transform>;
    kscan = <&ansi_kscan>;
};

&layout_60_iso {
    transform = <&iso_transform>;
    kscan = <&iso_kscan>;
};

&layout_60_all1u {
    transform = <&all_1u_transform>;
    kscan = <&all_1u_kscan>;
};

&layout_60_hhkb {
    transform = <&hhkb_transform>;
    kscan = <&hhkb_kscan>;
};
```

## Multiple Physical Layouts

> If a keyboard has multiple possible layouts (ex. you can snap off an outer column), then you should define multiple matrix transformations and multiple physical layouts - one for each possible layout.

Reviewer: ```suggestion
If a keyboard has multiple possible layouts (e.g. you can snap off an outer column), then you should define multiple matrix transformations and multiple physical layouts -- one for each possible layout.
```

If necessary, you can also define multiple kscan instances.

```dts
// Needed if and only if keys property is used
#include <physical_layouts.dtsi>

/ {
    default_layout: default_layout {
        compatible = "zmk,physical-layout";
        display-name = "Default Layout";
        transform = <&default_transform>;
        kscan = <&kscan0>;

>         keys = // List of key positions, optional

Reviewer: ```suggestion
        keys = <...>; // List of key positions, optional
```

    };

    alt_layout: alt_layout {
        compatible = "zmk,physical-layout";
        display-name = "Alternate Layout";
        transform = <&alt_transform>;
        kscan = <&alt_kscan0>;

>         keys = // List of key positions, optional

Reviewer: ```suggestion
        keys = <...>; // List of key positions, optional
```

    };
};
```

### Position Map

> A position map is used for precise mapping between layouts. It is used to allow [ZMK Studio](../../features/studio.md) to accurately retain keymaps when switching layouts. A keyboard's position map has a child node for every potential layout of the keyboard, with the corresponding entries in `positions` property will be used to determine the mapping between layouts. A position map looks something like this:

Reviewer: Is this missing an example after the last sentence?

Author: Yes, woops. Fixed.

        // Child node 1 here

        // Child node 2 here

        // ...
    };
};
```

Next, select a layout to be the "reference" layout. It is recommended that this be the layout which has the highest number of keys, as that prevents key mappings from being "lost" when switching back and forth between layouts.

Create the child node for the "reference" layout, and fill the `positions` array by iterating through the keys in your layout in the same order as the matrix transform. For a 2x2 macropad the child node would be

```dts
macropad {
    physical-layout = <&macropad_layout>;
    positions // This is equivalent to `positions = <0 1 2 3>;`, reshaped for readability
        = < 0  1 >
        , < 2  3 >;
};
```

Next, you'll need to write child nodes for every other layout, deriving the position map from the reference. Start by deleting any keys which aren't present in your layout, then reassign the numbers so that they match the matrix transform. Finally, (if the reference layout has more keys) add numbers back into the "empty"/"spare"/"nonexistent" positions so that both position maps count up to the same number. The below examples show this process more clearly.

> #### Example larger position map

Reviewer: Would it be worth adding an even more basic example as the first one that is not complete?

Author: Personally I'm not a fan of non-`complete;` position maps, I think they're more likely to lead to confusion. I'd rather omit in favor of encouraging people to be explicit.

Reviewer 2: Both are valid. iIMHO, and should be documented. Non-complete maps can save a ton of duplication, and not waste flash space for constrained devices. They *do* have a place, and should be documented.

                , <16 17 18 19>;
        };

        numpad {
            physical-layout = <&numpad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10 17>
                , <11 12 13 14>
                , <15 18 16 19>;
        };
    };
};
```

The "missing" positions are filled with the "spare" numbers of the layout with more keys. Meanwhile, if the right side with fewer keys is used as a reference, then the overall position map would look like this:

```dts
/ {
    keypad_lossy_position_map {
        compatible = "zmk,physical-layout-position-map";
        complete;

        macropad {
            physical-layout = <&macropad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10   >
                , <12 13 14 15>
                , <16    18   >;
        };

        numpad {
            physical-layout = <&numpad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10   >
                , <11 12 13 14>
                , <15    16   >;
        };
    };
};
```

> 

Reviewer: Is it worth adding a note here on why the former is preferable to this? i.e. that the extra positions can be recovered

Author: Added a note.

:::

#### Writing a position map

Start by creating the parent node defining the position map:

```dts
/ {
    keypad_lossless_position_map {
        compatible = "zmk,physical-layout-position-map";
        complete; // Not necessary, but generally recommended if you use position maps

        // Child node 1 here

        // Child node 2 here

        // ...
    };
};
```

Next, select a layout to be the "reference" layout. It is recommended that this be the layout which has the highest number of keys, as that prevents key mappings from being "lost" when switching back and forth between layouts.

Create the child node for the "reference" layout, and fill the `positions` array by iterating through the keys in your layout in the same order as the matrix transform. For a 2x2 macropad the child node would be

```dts
macropad_map: macropad {
    physical-layout = <&macropad_layout>;
    positions // This is equivalent to `positions = <0 1 2 3>;`, reshaped for readability
        = < 0  1 >
        , < 2  3 >;
};
```

> Next, you'll need to write child nodes for every other layout, deriving the position map from the reference. Start by deleting any keys which aren't present in your layout, then reassign the numbers so that they match the matrix transform. Finally, (if the reference layout has more keys) add numbers back into the "empty"/"spare"/"nonexistent" positions so that both position maps count up to the same number. The below examples show this process more clearly.

Reviewer 3: I don't think this makes any sense because we haven't explained what `positions` actually does yet. This is a pretty difficult concept to explain in general, and I can't come up with a way to explain it that I think is very simple, so I think we should try to get as many ideas for this as possible and then see which one people find easiest to understand.

I think of it as the `positions` properties in all the layouts being parallel arrays. The values at the same index in each layout indicate which key positions are the same in each layout. If I switch from layout A to layout B, then the binding at key position `A.positions[i]` in the original layout is transferred to key position `B.positions[i]` in the new layout for every index `i` in the position map. 

Key positions that are past the end of the keymap can also be used in order to store a binding so that it is not lost when switching between layouts with different numbers of keys.

A diagram would probably be very helpful to explaining this too. For example, consider layout A which has three keys and layout B which replaces the first two keys with a 2U key. If the position map is defined like this

```dts
// key positions for layout_a
// +---+---+---+
// | 0 | 1 | 2 |
// +---+---+---+
layout_a {
    ...
    positions = <0 1 2>;
};

// key positions for layout_b
// +-------+---+
// |   0   | 1 |
// +-------+---+
layout_b {
    ...
    positions = <0 2 1>;
};
```
Then this is how the key bindings move when switching between layouts A and B
```
+---+---+---+
+ 0 | 1 | 2 |
+---+---+---+
  ^   ^   ^
  |   |   |
  |   +-- | -------+
  v       v        v
+-------+---+    .....
|   0   | 1 |    . 2 .
+-------+---+    .....
```
Note that the key at position 1 in layout A is moved to an unused position 2 in layout B so that it can be restored when switching back to layout A.

#### Example larger position map

Consider the following macropad/numpad with two physical layouts:

![A 4x5 numpad/macropad](../../assets/hardware-integration/numpad.svg)

Let us first consider each side individually. The "reference" position map of the left side would look like this:

```dts
macropad_map: macropad {
    physical-layout = <&macropad_layout>;
    positions
        = < 0  1  2  3>
        , < 4  5  6  7>
        , < 8  9 10 11>
        , <12 13 14 15>
        , <16 17 18 19>;
};
```

Meanwhile, the "reference" position map of the right side with fewer keys would look like this:

```dts
numpad_map: numpad {
    physical-layout = <&numpad_layout>;
    positions
        = < 0  1  2  3>
        , < 4  5  6  7>
        , < 8  9 10   >
        , <11 12 13 14>
        , <15    16   >;
};
```

> As a reminder, the `positions` property is a one-dimensional array like the `keys` propertly, presented nicely through the use of whitespace.

Reviewer: ```suggestion
As a reminder, the `positions` property is a one-dimensional array like the `keys` property, formatted nicely through the use of whitespace and angle bracket groupings.
```

> As a reminder, the ordering of the physical layout and the position map is the same as in the matrix transform and the keymap (all are just one-dimensional arrays presented nicely through the use of whitespace).

Author: ```suggestion
As a reminder, the `positions` property is a one-dimensional array like the `keys` propertly, presented nicely through the use of whitespace.
```

        };

        numpad_map: numpad {
            physical-layout = <&numpad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10 17>
                , <11 12 13 14>
                , <15 18 16 19>;
        };
    };
};
```

The "missing" positions are filled with the "spare" numbers of the layout with more keys. Meanwhile, if the right side with fewer keys is used as a reference, then the overall position map would look like this:

```dts
/ {
    keypad_lossy_position_map {
        compatible = "zmk,physical-layout-position-map";
        complete;

        macropad_map: macropad {
            physical-layout = <&macropad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10   >
                , <12 13 14 15>
                , <16    18   >;
        };

        numpad_map: numpad {
            physical-layout = <&numpad_layout>;
            positions
                = < 0  1  2  3>
                , < 4  5  6  7>
                , < 8  9 10   >
                , <11 12 13 14>
                , <15    16   >;
        };
    };
};
```

> The above example is "lossy" because (unlike the "lossless" example) if a user switches from the macropad layout to the numpad layout _and then_ switches from the numpad layout back to the macropad layout, the assignments to the keys present but not listed in the macropad's map are lost.

Reviewer: ```suggestion
The above example is "lossy" because (unlike the previous "lossless" example) if a user switches from the macropad layout to the numpad layout _and then_ switches from the numpad layout back to the macropad layout, the assignments to the keys present but not listed in the macropad's map are lost.
```
The only place lossless was mentioned was in the node name, so this probably helps

#### Example non-`complete` position map

> Consider the above device again - most of the positions have identical `keys` properties. For example, the numpad's `15` key and the macropad's `16` key would have the same physical property, and be mapped to each other automatically. The keys whose mappings are unable to be determined automatically are those with different physical characteristics - the 2u high and 2u wide keys, and their corresponding 1u counterparts.

Reviewer: ```suggestion
Consider the above device again -- most of the positions have identical `keys` properties. For example, the macropad's `0` key and the numpad's `0` key would have the same physical property, and be mapped to each other automatically. The keys whose mappings are unable to be determined automatically are those with different physical characteristics: the 2u high and 2u wide keys, and their corresponding 1u counterparts.
```
While I believe the previous version was technically true, the width of 16/15 were actually different so it had to be qualified. This is probably a simpler example. (also switched order to obey left/right)

A non-`complete` position map can be used to assign mappings to only these particular keys:

```dts
/ {
    keypad_lossy_position_map {
        compatible = "zmk,physical-layout-position-map";

        macropad_map: macropad {
            physical-layout = <&macropad_layout>;
            positions = <7 15 16>;
        };

        numpad_map: numpad {
            physical-layout = <&numpad_layout>;
            positions = <7 14 15>;
        };
    };
};
```

This is noticably simpler to write, and can be a useful way of saving flash space for memory-constrained devices. The above is a "lossy" mapping, though. While "lossless" non-`complete` mappings are possible, they can be counter-intuitive enough that it may be easier to write the full position map instead.

> For completion, the equivalent "lossless" non-`complete` position map is shown below:

Reviewer: ```suggestion
For compleness, the equivalent "lossless" non-`complete` position map is shown below:
```

:::

> The `keys` property is required for [ZMK Studio](../../features/studio.md) support. It is used to describe the physical attributes of each key position present in that layout. If this property is used, then you should define the physical layouts and any [position maps](#position-map) in a file called `<your keyboard>-layouts.dtsi`. This file should then be imported by the appropriate file (A `.overlay`, `.dts`, or a `.dtsi` which is itself imported by one of the previous).

Reviewer: ```suggestion
The `keys` property is required for [ZMK Studio](../../features/studio.md) support. It is used to describe the physical attributes of each key position present in that layout. If this property is used, then you should define the physical layouts and any [position maps](#position-map) in a file called `<your keyboard>-layouts.dtsi`. This file should then be imported by the appropriate file, such as an `.overlay`, `.dts`, or a `.dtsi` (last of which is itself imported by one of the previous).
```
